doc/neps/nep-0022-ndarray-duck-typing-overview.rst

```diff
@@ -0,0 +1,340 @@
+array interface include: dask, sparse arrays, masked arrays, arrays
+with units (unless they switch to using dtypes), labeled arrays,
+masked arrays, and so forth. Clear use cases lead to good and relevant
+APIs.
+
+Second, the Anna Karenina principle applies here: full duck arrays are
+all alike, but every partial duck array is partial in its own way:
+
+* ``xarray.DataArray`` is mostly a duck array, but has incompatible
+  broadcasting semantics.
+* ``xarray.Dataset`` wraps multiple arrays in one object; it still
+  implements some array interfaces like ``__array_ufunc__``, but
+  certainly not all of them.
+* ``pandas.Series`` has methods with similar behavior to numpy, but
+  unique null-skipping behavior.
+* scipy’s ``LinearOperator``\s support matrix multiplication and nothing else
+* h5py and similar libraries for accessing array storage have objects
+  that support numpy-like slicing and conversion into a full array,
+  but not computation.
+* Some classes may be similar to ndarray, but without supporting the
+  full indexing semantics.
+
+And so forth.
+
+Despite our best attempts, we haven't found any clear, unique way of
+slicing up the ndarray API into a hierarchy of related types that
+captures these distinctions; in fact, it’s unlikely that any single
+person even understands all the distinctions. And this is important,
+because we have a *lot* of APIs that we need to add duck array support
+to (both in numpy and in all the projects that depend on numpy!). By
+definition, these already work for ``ndarray``, so hopefully getting
+them to work for full duck arrays shouldn’t be so hard, since by
+definition full duck arrays act like ``ndarray``. It’d be very
+cumbersome to have to go through each function and identify the exact
+subset of the ndarray API that it needs, then figure out which partial
+array types can/should support it. Once we have things working for
+full duck arrays, we can go back later and refine the APIs needed
+further as needed. Focusing on full duck arrays allows us to start
+making progress immediately.
+
+In the future, it might be useful to identify specific use cases for
+duck arrays and standardize narrower interfaces targeted just at those
+use cases. For example, it might make sense to have a standard “array
+loader” interface that file access libraries like h5py, netcdf, pydap,
+zarr, ... all implement, to make it easy to switch between these
+libraries. But that’s something that we can do as we go, and it
+doesn’t necessarily have to involve the NumPy devs at all. For an
+example of what this might look like, see the documentation for
+`dask.array.from_array
+<http://dask.pydata.org/en/latest/array-api.html#dask.array.from_array>`__.
+
+
+Principle 2: Take advantage of duck typing
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+``ndarray`` has a very large API surface area::
+
+    In [1]: len(set(dir(np.ndarray)) - set(dir(object)))
+    Out[1]: 138
```

Comment:
67 of these are `__dunder__` methods, mostly due to the arithmetic operators. An easy way to exclude those is with:
```
>>> not_scalar_operators = set(dir(np.ndarray)) - set(dir(np.lib.mixins.NDArrayOperatorsMixin))
>>> len(not_scalar_operators)
96
```

Reply:
Indeed, but the point here is just that duck arrays have a lot of stuff they have to do, so I think those dunder methods are still appropriate. (Not that it even matters that much, as the text says this is a huge underestimate anyway.)